README.md
# Elasticsearch::Embedded

[![Gem Version](https://badge.fury.io/rb/elasticsearch-embedded.png)](http://badge.fury.io/rb/elasticsearch-embedded) [![Build Status](https://travis-ci.org/fabn/elasticsearch-embedded.svg?branch=master)](https://travis-ci.org/fabn/elasticsearch-embedded) [![Coverage Status](https://coveralls.io/repos/fabn/elasticsearch-embedded/badge.png)](https://coveralls.io/r/fabn/elasticsearch-embedded)

This gem allows to download and execute elasticsearch (single node or local cluster) within a local folder.

It also provides some utilities for usage in RSpec integration tests.

## Installation

Add this line to your application's Gemfile:

    gem 'elasticsearch-embedded'

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install elasticsearch-embedded

## Usage

### Standalone mode

After gem installation you can run `embedded-elasticsearch` executable, it accepts some options for cluster configuration

```
$ embedded-elasticsearch -h
Usage: embedded-elasticsearch [options]
    -w, --working-dir=WORKING_DIR    Elasticsearch working directory (default: `Dir.tmpdir` or `Rails.root.join("tmp")` within rails applications)
    -p, --port=PORT                  Port on which to run elasticsearch (default: 9250)
    -c, --cluster-name=NAME          Cluster name (default: elasticsearch_test)
    -n, --nodes=NODES                Number of nodes started in the cluster (default: 1)
        --timeout=TIMEOUT            Timeout when starting the cluster (default: 30)
    -l, --log-level=LEVEL            Logger verbosity, numbers allowed (1..5) or level names (debug, info, warn, error, fatal)
    -q, --quiet                      Disable stdout logging
    -S, --show-es-output             Enable elasticsearch output in stdout
    -V VERSION                       Elasticsearch version to use (default 1.2.1)
    -P                               Configure cluster to persist data across restarts
    -h, --help                       Show this message
    -v, --version                    Show gem version
```

In order to start a single node cluster (with in memory indices) just run

```
$ embedded-elasticsearch -w tmp
Starting ES 1.2.1 cluster with working directory set to /Users/fabio/work/elasticsearch-embedded/tmp. Process pid is 57245
Downloading elasticsearch 1.2.1 |                                                    ᗧ| 100% (648 KB/sec) Time: 00:00:34
Starting 1 Elasticsearch nodes........
--------------------------------------------------------------------------------
Cluster:            elasticsearch_test
Status:             green
Nodes:              1
                    + node-1 | version: 1.2.1, pid: 57254, address: inet[/0:0:0:0:0:0:0:0:9250]

# Your cluster is running and listening to port 9250
```

### Usage with foreman

```
$ cat Procfile
elasticsearch: embedded-elasticsearch -w tmp
$ foreman start
14:53:51 elasticsearch.1 | started with pid 57524
14:53:51 elasticsearch.1 | Starting ES 1.2.1 cluster with working directory set to /Users/fabionapoleoni/Desktop/work/RubyMine/elasticsearch-embedded/tmp. Process pid is 57524
14:53:57 elasticsearch.1 | Starting 1 Elasticsearch nodes........
14:53:57 elasticsearch.1 | --------------------------------------------------------------------------------
14:53:57 elasticsearch.1 | Cluster:            elasticsearch_test
14:53:57 elasticsearch.1 | Status:             green
14:53:57 elasticsearch.1 | Nodes:              1
14:53:57 elasticsearch.1 |                     + node-1 | version: 1.2.1, pid: 57528, address: inet[/0:0:0:0:0:0:0:0:9250]
^CSIGINT received
14:54:02 system          | sending SIGTERM to all processes
14:54:02 elasticsearch.1 | exited with code 0%
```

### With RSpec

```ruby
# In spec/spec_helper.rb
require 'elasticsearch-embedded'
# Activate gem behavior with :elasticsearch tagged specs
Elasticsearch::Embedded::RSpec.configure
# Alternatively you could specify for which tags you want gem support
Elasticsearch::Embedded::RSpec.configure_with :search

# In tagged specs
describe Something, :elasticsearch do

  before(:each) do
    # all indices are deleted automatically at beginning of each spec
    expect(client.indices.get_settings).to be_empty
  end

  # If elastic search client is defined (i.e. with gem 'elasticsearch' or require 'elasticsearch')
  it 'should make elastic search client available' do
    expect(client).to be_an_instance_of(::Elasticsearch::Transport::Client)
    # Create a document into test cluster
    client.index index: 'test', type: 'test-type', id: 1, body: {title: 'Test'}
  end

  # Cluster object is also exposed in a helper
  it 'should make cluster object available' do
    expect(cluster).to be_an_instance_of(::Elasticsearch::Embedded::Cluster)
  end

  # If elastic search client is not defined client return an URI instance with base url for the cluster
  it 'should return cluster uri when elasticsearch is not defined' do
    ::Elasticsearch.send(:remove_const, :Client)
    expect(client).to be_an_instance_of(::URI::HTTP)
    expect(Net::HTTP.get(client)).to include('You Know, for Search')
  end

end

```

### With VCR/Webmock

If your test suite include a web stubbing library this gem will have issues, for instance VRC and/or Webmock disable
web interactions making impossible to download ES (on first run) and to connect to the cluster. Temporarily disable
them with around hooks, in RSpec

```ruby
# Disable VCR and webmock for elasticsearch tagged specs
config.around(:each, :elasticsearch) do |example|
  begin
    WebMock.allow_net_connect!
    VCR.turned_off(&example)
  ensure
    WebMock.disable_net_connect!
  end
end
```

### With Test::Unit/minitest

Pull requests are welcome.

## Contributing

1. Fork it ( https://github.com/fabn/elasticsearch-embedded/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


Changelog.md
## On develop branch

## 0.1.2, released 2015-11-11

* Prevent warnings when used with RSpec 2.x.
* Avoid name issues when `elasticsearch-rails` gem is present in the bundle

## 0.1.1, released 2014-11-15

* Updated elasticsearch default version to 1.4.0
* Relaxed runtime dependencies versions (from `~> x.y.z` to `~> x.y`)

## 0.1.0, released 2014-06-25

* Initial version


spec/spec_helper.rb
require 'coveralls'
Coveralls.wear!
# Main gem file
require 'elasticsearch-embedded'
# Used to stub calls to filesystem
require 'fakefs/spec_helpers'

# Used in integration tests
require 'elasticsearch'
# Activate gem behavior with :elasticsearch tagged specs
Elasticsearch::Embedded::RSpec.configure
# disable logging in tests
Logging.logger.root.clear_appenders

RSpec.configure do |config|
  config.filter_run :focus
  config.run_all_when_everything_filtered = true

  # Many RSpec users commonly either run the entire suite or an individual
  # file, and it's useful to allow more verbose output when running an
  # individual spec file.
  if config.files_to_run.one?
    # Use the documentation formatter for detailed output,
    # unless a formatter has already been configured
    # (e.g. via a command-line flag).
    config.default_formatter = 'doc'
  end

  # Print the 10 slowest examples and example groups at the
  # end of the spec run, to help surface which specs are running
  # particularly slow.
  config.profile_examples = 10

  # Run specs in random order to surface order dependencies. If you find an
  # order dependency and want to debug it, you can fix the order by providing
  # the seed, which is printed after each run.
  #     --seed 1234
  config.order = :random

  # Seed global randomization in this process using the `--seed` CLI option.
  # Setting this allows you to use `--seed` to deterministically reproduce
  # test failures related to randomization by passing the same `--seed` value
  # as the one that triggered the failure.
  Kernel.srand config.seed

  # rspec-expectations config goes here. You can use an alternate
  # assertion/expectation library such as wrong or the stdlib/minitest
  # assertions if you prefer.
  config.expect_with :rspec do |expectations|
    # Enable only the newer, non-monkey-patching expect syntax.
    # For more details, see:
    #   - http://myronmars.to/n/dev-blog/2012/06/rspecs-new-expectation-syntax
    expectations.syntax = :expect
  end

  # rspec-mocks config goes here. You can use an alternate test double
  # library (such as bogus or mocha) by changing the `mock_with` option here.
  config.mock_with :rspec do |mocks|
    # Enable only the newer, non-monkey-patching expect syntax.
    # For more details, see:
    #   - http://teaisaweso.me/blog/2013/05/27/rspecs-new-message-expectation-syntax/
    mocks.syntax = :expect

    # Prevents you from mocking or stubbing a method that does not exist on
    # a real object. This is generally recommended.
    mocks.verify_partial_doubles = true
  end
end


spec/lib/elasticsearch/embedded_spec.rb
describe Elasticsearch::Embedded do

  describe '.verbosity' do

    after { Logging.logger[subject].level = :info }

    it 'should allow numbers' do
      expect {
        subject.verbosity(2)
      }.to change { Logging.logger[subject].level }.from(Logging::LEVELS['info']).to(Logging::LEVELS['warn'])
    end

    it 'should allow level strings' do
      expect {
        subject.verbosity('warn')
      }.to change { Logging.logger[subject].level }.from(Logging::LEVELS['info']).to(Logging::LEVELS['warn'])
    end

    it 'should keep level to info on invalid levels' do
      expect {
        subject.verbosity('foo')
      }.to_not change { Logging.logger[subject].level }
    end

  end

  describe '.mute!' do

    after { Logging.logger[subject].appenders = Logging.appenders.stdout }

    it 'should remove all appenders' do
      expect {
        subject.mute!
      }.to change { Logging.logger[subject].appenders }.to([])
    end

  end

end

spec/lib/elasticsearch/embedded/cluster_spec.rb
describe Elasticsearch::Embedded::Cluster, :elasticsearch do

  describe '#clear_all_indices!' do

    it 'should delete all indices' do
      # Create a document and an index
      client.index index: 'test', type: 'test-type', id: 1, body: {title: 'Test'}
      expect {
        cluster.delete_all_indices!
      }.to change { client.indices.get_settings.count }.to(0)
    end

    it 'should delete all indices' do
      # Create a document and an index
      client.index index: 'test', type: 'test-type', id: 1, body: {title: 'Test'}
      client.index index: 'test2', type: 'test-type', id: 1, body: {title: 'Test'}
      expect {
        cluster.delete_index! 'test'
      }.to change { client.indices.get_settings.keys }.to(['test2'])
    end

  end

  describe '#pids' do

    it 'should return a list running instances pids' do
      expect(cluster.pids).to_not be_empty
    end

    it 'should return an empty array when not started' do
      not_started = Elasticsearch::Embedded::Cluster.new
      not_started.port = 50000 # Likely there's no process running on this port
      expect(not_started.pids).to eq([])
    end

  end

  describe 'Development template' do

    before do
      # It's not applied by default on non persistent clusters
      cluster.apply_development_template!
      client.indices.create index: 'any_index'
    end

    let(:index_settings) { client.indices.get_settings(index: 'any_index')['any_index']['settings']['index'] }

    # Do not leave status across tests
    after(:all) do
      client.indices.delete_template name: 'development_template' if client.indices.get_template.has_key?('development_template')
    end

    it 'should configure 1 shard for each index' do
      expect(index_settings).to include('number_of_shards' => '1')
    end

    it 'should configure 0 replicas for each index' do
      expect(index_settings).to include('number_of_replicas' => '0')
    end

  end

  describe 'Cluster persistency' do

    let(:persistent_cluster) { Elasticsearch::Embedded::Cluster.new }
    let(:port) { cluster.port + 10 }
    let(:client) { Elasticsearch::Client.new host: "localhost:#{port}" }

    before do
      persistent_cluster.persistent = true
      persistent_cluster.cluster_name = 'persistent_cluster'
      persistent_cluster.port = port
    end

    after do
      # Ensure additional cluster is stopped when test is finished
      persistent_cluster.stop if persistent_cluster.running?
    end

    it 'should persist data across restarts' do
      persistent_cluster.start
      # Index a document and trigger persistent index creation
      client.index index: 'persistent', type: 'test-type', id: 1, body: {title: 'Test'}, refresh: true
      # Restart cluster and check index presence
      expect {
        persistent_cluster.stop
        persistent_cluster.start
      }.to_not change { client.indices.get_settings }
    end

  end

end

lib/elasticsearch/embedded/downloader.rb
# coding: utf-8
require 'tmpdir'
require 'open-uri'
require 'ruby-progressbar'
require 'zip'

module Elasticsearch
  module Embedded
    class Downloader

      # Default temporary path used by downloader
      TEMPORARY_PATH = defined?(::Rails) ? ::Rails.root.join('tmp') : Dir.tmpdir
      # Default version of elasticsearch to download
      DEFAULT_VERSION = '1.4.0'

      attr_reader :version, :path

      def initialize(args = {})
        @version = args[:version] || ENV['ELASTICSEARCH_VERSION'] || DEFAULT_VERSION
        @path = args[:path] || ENV['ELASTICSEARCH_DOWNLOAD_PATH'] || TEMPORARY_PATH
      end

      # Download elasticsearch distribution and unzip it in the specified temporary path
      def self.download(arguments = {})
        new(arguments).perform
      end

      def perform
        download_file
        extract_file
        self
      end

      def downloaded?
        File.exists?(final_path)
      end

      def extracted?
        File.directory?(dist_folder)
      end

      def dist_folder
        @dist_folder ||= final_path.gsub /\.zip\Z/, ''
      end

      def final_path
        @final_path ||= File.join(working_dir, "elasticsearch-#{version}.zip")
      end

      def working_dir
        @working_dir ||= File.realpath(path)
      end

      def executable
        @executable ||= File.join(dist_folder, 'bin', 'elasticsearch')
      end

      private

      def download_file
        return if downloaded?
        open(final_path, 'wb') do |target|
          download_options = {
              content_length_proc: ->(t) { build_progress_bar(t) },
              progress_proc: ->(s) { increment_progress(s) }
          }
          # direct call here to avoid spec issues with Kernel#open
          distfile = OpenURI.open_uri("https://download.elasticsearch.org/elasticsearch/elasticsearch/elasticsearch-#{version}.zip", download_options)
          target << distfile.read
        end
      end

      def extract_file
        return if extracted?
        Dir.chdir(working_dir) do
          # Extract archive in path, after block CWD is restored
          Zip::File.open(final_path) do |zip_file|
            # Extract all entries into working dir
            zip_file.each(&:extract)
          end
        end
        # ensure main executable has execute permission
        File.chmod(0755, executable)
        # Create folder for log files
        FileUtils.mkdir(File.join(dist_folder, 'logs'))
      end

      # Build a progress bar to download elasticsearch
      def build_progress_bar(total)
        if total && total.to_i > 0
          @download_progress_bar = ProgressBar.create title: "Downloading elasticsearch #{version}", total: total,
                                                      format: '%t |%bᗧ%i| %p%% (%r KB/sec) %e', progress_mark: ' ', remainder_mark: '･',
                                                      rate_scale: ->(rate) { rate / 1024 }, smoothing: 0.7
        end
      end

      def increment_progress(size)
        @download_progress_bar.progress = size if @download_progress_bar
      end

    end

  end
end


lib/elasticsearch/embedded/rspec_configuration.rb
module Elasticsearch
  module Embedded

    module RSpec

      module ElasticSearchHelpers
        # Return a client connected to the configured client, if ::Elasticsearch::Client is defined
        # return a client, else return a URI attached to cluster.
        # @see http://ruby-doc.org/stdlib-2.1.2/libdoc/net/http/rdoc/Net/HTTP.html
        # @see https://github.com/elasticsearch/elasticsearch-ruby
        def client
          @client ||=
              case
                when defined?(::Elasticsearch::Client)
                  ::Elasticsearch::Client.new host: "localhost:#{cluster.port}"
                else
                  URI("http://localhost:#{cluster.port}/")
              end
        end

        # Return a cluster instance to be used in tests
        def cluster
          ElasticSearchHelpers.memoized_cluster
        end

        class << self

          # Return a singleton instance of cluster object
          def memoized_cluster
            @cluster ||= ::Elasticsearch::Embedded::Cluster.new
          end

        end

      end

      class << self

        # Configure rspec for usage with ES cluster
        def configure_with(*meta)
          # assign default value to tags
          ::RSpec.configure do |config|

            # Include helpers only in tagged specs
            config.include ElasticSearchHelpers, *meta

            # Before hook, starts the cluster
            config.before(:each, *meta) do
              ElasticSearchHelpers.memoized_cluster.ensure_started!
              ElasticSearchHelpers.memoized_cluster.delete_all_indices!
            end

            # After suite hook, stop the cluster
            config.after(:suite) do
              ElasticSearchHelpers.memoized_cluster.stop if ElasticSearchHelpers.memoized_cluster.running?
            end
          end
        end

        # Default config method, configure RSpec with :elasticsearch filter only.
        # Equivalent to .configure_with(:elasticsearch)
        def configure
          configure_with(elasticsearch: true)
        end
      end

    end

  end
end


lib/elasticsearch/embedded/logger_configuration.rb
require 'logging'

module Elasticsearch
  module Embedded

    # Contains stuff related to logging configuration
    module LoggerConfiguration

      # Configure logger verbosity for ::Elasticsearch::Embedded log hierarchy
      # @param [String,Fixnum] level accepts strings levels or numbers
      def verbosity(level)
        level = level.to_s.downcase # normalize string to downcase
        case
          when level =~ /\A\d\Z/
            Logging.logger[self].level = level.to_i
          when Logging::LEVELS.include?(level)
            Logging.logger[self].level = Logging::LEVELS[level]
          else
            Logging.logger[self].level = :info
        end

      end

      # Clear all logging appenders for ::Elasticsearch::Embedded log hierarchy
      def mute!
        Logging.logger[self].clear_appenders
      end

      # @see https://github.com/TwP/logging/blob/master/examples/colorization.rb
      def configure_logging!
        # Configure logger levels for hierarchy
        Logging.logger[self].level = :info
        # Register a color scheme named bright
        Logging.color_scheme 'bright', {
            levels: {
                info: :green,
                warn: :yellow,
                error: :red,
                fatal: [:white, :on_red]
            },
            date: :blue,
            logger: :cyan,
            message: :white,
        }
        pattern_options = {pattern: '[%d] %-5l %c: %m\n'}
        # Apply colors only if in tty
        pattern_options[:color_scheme] = 'bright' if STDOUT.tty?
        # Create a named appender to be used only for current module
        Logging.logger[self].appenders = Logging.appenders.stdout(self.to_s, layout: Logging.layouts.pattern(pattern_options))
      end

      private

      # Extension callback
      def self.extended(base)
        base.configure_logging!
      end

    end

  end
end


lib/elasticsearch/embedded/cluster.rb
require 'timeout'
require 'net/http'
require 'uri'
require 'json'

module Elasticsearch
  module Embedded

    # Class used to manage a local cluster of elasticsearch nodes
    class Cluster

      # Make logger method available
      include Logging.globally

      # Options for cluster
      attr_accessor :port, :cluster_name, :nodes, :timeout, :persistent, :additional_options, :verbose

      # Options for downloader
      attr_accessor :downloader, :version, :working_dir

      # Assign default values to options
      def initialize
        @nodes = 1
        @port = 9250
        @version = Downloader::DEFAULT_VERSION
        @working_dir = Downloader::TEMPORARY_PATH
        @timeout = 30
        @cluster_name = 'elasticsearch_test'
        @pids = []
        @pids_lock = Mutex.new
      end

      # Start an elasticsearch cluster and return immediately
      def start
        @downloader = Downloader.download(version: version, path: working_dir)
        start_cluster
        apply_development_template! if persistent
      end

      # Start an elasticsearch cluster and wait until running, also register
      # a signal handler to close the cluster on INT, TERM and QUIT signals
      def start_and_wait!
        # register handler before starting cluster
        register_shutdown_handler
        # start the cluster
        start
        # Wait for all child processes to end then return
        Process.waitall
      end

      # Stop the cluster and return after all child processes are dead
      def stop
        logger.warn 'Cluster is still starting, wait until startup is complete before sending shutdown command' if @pids_lock.locked?
        @pids_lock.synchronize do
          http_object.post('/_shutdown', nil)
          logger.debug 'Cluster stopped succesfully using shutdown api'
          Timeout.timeout(2) { Process.waitall }
          # Reset running pids reader
          @pids = []
        end
      rescue
        logger.warn "Following processes are still alive #{pids}, kill them with signals"
        # Send term signal if post request fails to all processes still alive after 2 seconds
        pids.each { |pid| wait_or_kill(pid) }
      end

      # Thread safe access to all spawned process pids
      def pids
        @pids_lock.synchronize { @pids }
      end

      # Start server unless it's running
      def ensure_started!
        start unless running?
      end

      # Returns true when started cluster is running
      #
      # @return Boolean
      def running?
        cluster_health = Timeout::timeout(0.25) { __get_cluster_health } rescue nil
        # Response is present, cluster name is the same and number of nodes is the same
        !!cluster_health && cluster_health['cluster_name'] == cluster_name && cluster_health['number_of_nodes'] == nodes
      end

      # Remove all indices in the cluster
      #
      # @return [Array<Net::HTTPResponse>] raw http responses
      def delete_all_indices!
        delete_index! :_all
      end

      # Remove the indices given as args
      #
      # @param [Array<String,Symbol>] args list of indices to delet
      # @return [Array<Net::HTTPResponse>] raw http responses
      def delete_index!(*args)
        args.map { |index| http_object.request(Net::HTTP::Delete.new("/#{index}")) }
      end

      # Used for persistent clusters, otherwise cluster won't get green state because of missing replicas
      def apply_development_template!
        development_settings = {
            template: '*',
            settings: {
                number_of_shards: 1,
                number_of_replicas: 0,
            }
        }
        # Create the template on cluster
        http_object.put('/_template/development_template', JSON.dump(development_settings))
      end

      private

      # Build command line to launch an instance
      def build_command_line(instance_number)
        [
            downloader.executable,
            '-D es.foreground=yes',
            "-D es.cluster.name=#{cluster_name}",
            "-D es.node.name=node-#{instance_number}",
            "-D es.http.port=#{port + (instance_number - 1)}",
            "-D es.gateway.type=#{cluster_options[:gateway_type]}",
            "-D es.index.store.type=#{cluster_options[:index_store]}",
            "-D es.path.data=#{cluster_options[:path_data]}-#{instance_number}",
            "-D es.path.work=#{cluster_options[:path_work]}-#{instance_number}",
            '-D es.network.host=0.0.0.0',
            '-D es.discovery.zen.ping.multicast.enabled=true',
            '-D es.script.disable_dynamic=false',
            '-D es.node.test=true',
            '-D es.node.bench=true',
            additional_options,
            verbose ? nil : '> /dev/null'
        ].compact.join(' ')
      end

      # Spawn an elasticsearch process and return its pid
      def launch_instance(instance_number = 1)
        # Start the process within a new process group to avoid signal propagation
        Process.spawn(build_command_line(instance_number), pgroup: true).tap do |pid|
          logger.debug "Launched elasticsearch process with pid #{pid}, detaching it"
          Process.detach pid
        end
      end

      # Return running instances pids, borrowed from code in Elasticsearch::Extensions::Test::Cluster.
      # This method returns elasticsearch nodes pids and not spawned command pids, they are different because of
      # elasticsearch shell wrapper used to launch the daemon
      def nodes_pids
        # Try to fetch node info from running cluster
        nodes = JSON.parse(http_object.get('/_nodes/?process').body) rescue []
        # Fetch pids from returned data
        nodes.empty? ? nodes : nodes['nodes'].map { |_, info| info['process']['id'] }
      end

      def start_cluster
        logger.info "Starting ES #{version} cluster with working directory set to #{working_dir}. Process pid is #{$$}"
        if running?
          logger.warn "Elasticsearch cluster already running on port #{port}"
          wait_for_green(timeout)
          return
        end
        # Launch single node instances of elasticsearch with synchronization
        @pids_lock.synchronize do
          1.upto(nodes).each do |i|
            @pids << launch_instance(i)
          end
          # Wait for cluster green state before releasing lock
          wait_for_green(timeout)
          # Add started nodes pids to pid array
          @pids.concat(nodes_pids)
        end
      end

      def wait_or_kill(pid)
        begin
          Timeout::timeout(2) do
            Process.kill(:TERM, pid)
            logger.debug "Sent SIGTERM to process #{pid}"
            Process.waitpid(pid)
            logger.info "Process #{pid} exited successfully"
          end
        rescue Errno::ESRCH, Errno::ECHILD
          # No such process or no child => process is already dead
          logger.debug "Process with pid #{pid} is already dead"
        rescue Timeout::Error
          logger.info "Process #{pid} still running after 2 seconds, sending SIGKILL to it"
          Process.kill(:KILL, pid) rescue nil
        ensure
          logger.debug "Removing #{pid} from running pids"
          @pids_lock.synchronize { @pids.delete(pid) }
        end
      end

      # Used as arguments for building command line to launch elasticsearch
      def cluster_options
        {
            port: port,
            nodes: nodes,
            cluster_name: cluster_name,
            timeout: timeout,
            # command to run is taken from downloader object
            command: downloader.executable,
            # persistency options
            gateway_type: persistent ? 'local' : 'none',
            index_store: persistent ? 'mmapfs' : 'memory',
            path_data: File.join(persistent ? downloader.working_dir : Dir.tmpdir, 'cluster_data'),
            path_work: File.join(persistent ? downloader.working_dir : Dir.tmpdir, 'cluster_workdir'),
        }
      end

      # Return an http object to make requests
      def http_object
        @http ||= Net::HTTP.new('localhost', port)
      end

      # Register a shutdown proc which handles INT, TERM and QUIT signals
      def register_shutdown_handler
        stopper = ->(sig) do
          Thread.new do
            logger.info "Received SIG#{Signal.signame(sig)}, quitting"
            stop
          end
        end
        # Stop cluster on Ctrl+C, TERM (foreman) or QUIT (other)
        [:TERM, :INT, :QUIT].each { |sig| Signal.trap(sig, &stopper) }
      end

      # Waits until the cluster is green and prints information
      #
      # @example Print the information about the default cluster
      #     Elasticsearch::Extensions::Test::Cluster.wait_for_green
      #
      # @param (see #__wait_for_status)
      #
      # @return Boolean
      #
      def wait_for_green(timeout = 60)
        __wait_for_status('green', timeout)
      end

      # Blocks the process and waits for the cluster to be in a "green" state.
      #
      # Prints information about the cluster on STDOUT if the cluster is available.
      #
      # @param status  [String]  The status to wait for (yellow, green)
      # @param timeout [Integer] The explicit timeout for the operation
      #
      # @api private
      #
      # @return Boolean
      #
      def __wait_for_status(status='green', timeout = 30)
        Timeout::timeout(timeout) do
          loop do
            response = JSON.parse(http_object.get("/_cluster/health?wait_for_status=#{status}").body) rescue {}

            # check response and return if ok
            if response['status'] == status && nodes == response['number_of_nodes'].to_i
              __print_cluster_info and break
            end

            logger.debug "Still waiting for #{status} status in #{cluster_name}"
            sleep 1
          end
        end

        true
      end

      # Print information about the cluster on STDOUT
      #
      # @api private
      #
      def __print_cluster_info
        health = JSON.parse(http_object.get('/_cluster/health').body)
        nodes = JSON.parse(http_object.get('/_nodes/process,http').body)
        master = JSON.parse(http_object.get('/_cluster/state').body)['master_node']

        logger.info '-'*80
        logger.info 'Cluster: '.ljust(12) + health['cluster_name'].to_s
        logger.info 'Status:  '.ljust(12) + health['status'].to_s
        logger.info 'Nodes:   '.ljust(12) + health['number_of_nodes'].to_s

        nodes['nodes'].each do |id, info|
          m = id == master ? '+' : '-'
          logger.info ''.ljust(12) + "#{m} #{info['name']} | version: #{info['version']}, pid: #{info['process']['id']}, address: #{info['http']['bound_address']}"
        end
      end

      # Tries to load cluster health information
      #
      # @api private
      #
      def __get_cluster_health
        JSON.parse(http_object.get('/_cluster/health').body) rescue nil
      end

    end

  end
end
